### ISSUE TITLE
Implicit conversions differ when using signed and unsigned integers

### ISSUE THREAD

**Issue by larsbohn** [ispc/ispc#2714]
I stumbled across this just the other day.

https://ispc.godbolt.org/z/PK5aYWsaa

```c
void main()
{
    float f = 0.1f;

    uint u32 = 1;
    uint64 u64 = 1;

    float result_u32 = f * u32;
    float result_u64 = f * u64;

    int i32 = 1;
    int64 i64 = 1;

    float result_i32 = f * i32;
    float result_i64 = f * i64;
}
```

So the behaviour is different based on whether uint32 or uint64 is being used for unsigned.

For signed it seems to yield the correct result, though not sure.

This is quite subtle and differs from how a regular C/C++ compiler handles this, although warnings are issued.

What do you think, should this be changed?

**Comment by nurmukhametov** [comment]
Yes, it is so as @dbabokin said here https://github.com/ispc/ispc/issues/740#issuecomment-1835759362.

Both `int64` and `uint64` are prioritized over `float`, i.e. with your example :

```bash
result_u32 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
result_u64 [0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000]
result_i32 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
result_i64 [0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000]
```

To workaround this, you may explicitly cast `uint64` to `float`: 

```ispc
float result_u64 = f * (float) u64;
````

to produce the expected result:

```bash
result_u32 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
result_u64 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
result_i32 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
result_i64 [0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000,0.100000]
````

I think it is better to change that as https://github.com/ispc/ispc/issues/740 suggests but I am slightly concerned how it may impact user code.

**Comment by pbrubaker** [comment]
We could issue a warning here.  Informing the developer of the current precedence if there isn't an explicit cast.  That's at least something we can do without breaking user code.

If we do change the behavior we could add a switch to disable the change and/or add a warning that the behavior has changed.

**Comment by pbrubaker** [comment]
We could improve the warning here and inform the user of the current precedence when an explicit cast isn't used - instead of just the performance warning.

`<source>:8:28: Performance Warning: Conversion from uint32 to float 
        is slow. Use "int32" if possible `

If we do change the behavior as mentioned in #740 I think we could mitigate the impact in a couple ways: a compiler switch to revert back to the original behavior, and a similar warning when an explicit cast is not present.

**Comment by nurmukhametov** [comment]
> We could improve the warning here and inform the user of the current precedence when an explicit cast isn't used - instead of just the performance warning.
> 
> `<source>:8:28: Performance Warning: Conversion from uint32 to float is slow. Use "int32" if possible `

Agree that we need warning to warn user about unexpected type priority (from the C/C++ point of view).

**Comment by larsbohn** [comment]
I think a default warning about precedence would be a good solution so we can explicitly insert casts where appropriate.

Else this will go unnoticed when porting code to ISPC.

**Comment by aneshlya** [comment]
@claude , implement the proposed warning.
